Revert changes to DataContainer storage format

pmrv opened the pull request:
PR #1364 changed saved the contents of the container in a data subgroup, but that breaks backwards compat and is not actually necessary for the TemplateJob, so revert it here.

pmrv committed:
Revert changes to DataContainer storage format

PR #1364 changed saved the contents of the container in a data subgroup,
but that breaks backwards compat and is not actually necessary for the
TemplateJob, so revert it here.

pmrv committed:
Handle list in to_dict

jan-janssen commented:
It is only the new test that fails:
```
======================================================================
FAIL: test_dict_empty_list (generic.test_datacontainer.TestDataContainer.test_dict_empty_list) [to/from_dict should give back the same list as written.]
HasDict interface should work
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/pyiron_base/pyiron_base/tests/unit/generic/test_datacontainer.py", line 655, in test_dict_empty_list
    self.assertEqual(self.pl, l)
AssertionError: DataContainer({0: DataContainer({'foo': 'bar'}), 1: 2, 2: 4[110 chars]8])}) != DataContainer({0: {'foo': 'bar'}, 1: 2, 2: 42, 3: {'next': [35 chars] 8]})

----------------------------------------------------------------------
```

jan-janssen commented:
@pmrv While the tests in this pull request fail, this fixes the issue in the notebook tests of `pyiron_atomistics`.

pmrv commented:
I have a fix for the tests that I'll polish and push on Monday.

pmrv committed:
Add split_children_dict to undo join_children_dict

HasDict.to_dict uses join_children_dict to enable more efficient writing
to HDF.  When reading from HDF this does not matter, because h5io_browser
returns the dictionaries correctly, but when trying something like

>>> data = obj.to_dict()
>>> obj.from_dict(data)

objects can choke on the joined keys, so now from_dict splits the keys automatically
in case it finds any.

pmrv committed:
Save key order in DataContainer.to_dict

Also let the generic code from HasDict take care of the conversion of children.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jan-janssen committed:
Merge pull request #1623 from pyiron/main

Merge main

jan-janssen committed:
Update unittests_old.yml

jan-janssen committed:
Update pyproject.toml
